sapirbitan/assignment2/ex2.cpp

```diff
@@ -0,0 +1,104 @@
+#include <iostream>
+using namespace std;
+
+struct node {
+	int key;
+	node* left_child = NULL;
+	node* right_child = NULL;
+	node* ancestor = NULL;
+};//assuming every node contains pointers to it's children and ancestors
+  //(the root's ancestor will point to NULL)
+
+/****************************************************************************************
+This function accepts three nodes - the root of the tree, and the two nodes we want to find
+the common ancestor of.
+how does she do it?
+It goes over all of the second node's ancestor and compares them with all of the first node's 
+ancestors.
+Worst case scenario - both of the nodes are at the end of the tree - which meens we'll have 
+to compare log(n) [n is num of nodes in the tree] ancestors with log(n) ancestors - les then o(n)
+*****************************************************************************************/
+node* find_common_anc(const node& tree_root,const node& first_node, const node& second_node) {
+	//Since I don't want to change the original nodes(ans can't - because they're const)
+	//I am using temporary nodes which I can change
+	node const_second = second_node;
+	node temp_first = first_node;
+	node temp_second = second_node;
+	node temp_tree_root = tree_root;
+	int flag=1;//checks if we've found the common ancestor
+	//I wan't this loop to keep going until the ancestor is found - so while(1) ;)
+	while (flag) {
+		if ((first_node.key == tree_root.key) || (second_node.key == tree_root.key)){
+			cout << "one of the nodes is the root - which doesn't have an ancestor" << endl;
+			//edge case - if someone sends the root- it doesn't have an ancestor - because he is the ancestor
+			return &temp_tree_root;
+		}
+		if (temp_second.key == tree_root.key) {//if when we went over all of the ancestors
+			//of the second node and reached the root, we'll start the second iteration - 
+			//the first node we'll go up to it's ancestor, and the second we'll go back to it's original value
+			temp_first = *temp_first.ancestor;
+			temp_second = const_second;
+		}
+		//if it's not the same ancestor - go to the upper ancestor
+		else if (temp_first.ancestor->key != temp_second.ancestor->key)
+			temp_second = *temp_second.ancestor;
+		//else - it is the one we were looking for, return it.
+		else {
+			flag=0;
```

Comment:
This is not necessary :)

`flag` is used to control the while loop. But once you find the node you're looking for, you return it and exit out of the loop, so I think your code could be replaced by:

```
while(true) {
    if ((first_node.key == tree_root.key) || (second_node.key == tree_root.key)){...}

    if (temp_second.key == tree_root.key) {...}

    else if (temp_first.ancestor->key != temp_second.ancestor->key) {...}
   
    else return temp_first.ancestor;
}
```

Usually, `while(true)` can be pretty dangerous as you can get into an infinite loop. But I think in your case it's ok because you're always guaranteed to find a common ancestor of two nodes in a tree :)